### PR TITLE
Tests and multirole in QGIS plugin

### PR DESCRIPTION
**Pull request by arnaud-morvan**
* Add doctstrings and README.

* Add base tests in docker/qgisserver (not exhautive) for QGIS 3.4. Build a docker image for tests on top of production geomapfish QGIS docker image to run the tests.

* Support multi-role.

### CODE REVIEW

File: docker/build/requirements.txt [hunk 1]
```diff
@@ -15,6 +15,7 @@ coverage==4.5.2  # Build coverage XML for Codacy
 dateutils==0.6.6  # geoportal pcreate
 deform==2.0.7  # commons, admin
 defusedxml==0.5.0  # geoportal
+docker-compose==1.23.2  # qgisserver tests
```
[the remark this answers is not in the document]
Comment (arnaud-morvan) [reply docker/build/requirements.txt]: We could run:
```
./docker-run make tests 
make -C docker/qgisserver make tests
```
But here, I've added ```make -C docker/qgisserver make tests``` in root tests target that is run through ./docker-run.

File: docker/qgisserver/geomapfish_qgisserver/accesscontrol.py [hunk 2]
```diff
@@ -208,64 +216,94 @@ def assert_plugin_initialised(self):
...
         parameters = self.serverInterface().requestHandler().parameterMap()
 
-        if parameters.get('ROLE_ID') == "0" and parameters.get('USER_ID') == "0":
+        if parameters.get('USER_ID') == "0":
             return "ROOT"
 
-        role = self.DBSession.query(Role).get(parameters['ROLE_ID']) if 'ROLE_ID' in parameters else None
-        QgsMessageLog.logMessage("Role: {}".format(role.name if role else '-'))
-        return role
+        roles = set(
```
[the remark this answers is not in the document]
Comment (arnaud-morvan) [reply docker/qgisserver/geomapfish_qgisserver/accesscontrol.py]: To avoid duplicates.

**Comment by arnaud-morvan** [comment]
For me it is really easier to develop when running the tests using docker-compose.
Easy to use docker-compose.override.yaml to mount volumes.
No need to run linked containers manually.

File: docker/qgisserver/acceptance_tests/functional/conftest.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    #fileConfig(config_uri, defaults=dict(os.environ))
+    #settings = plaster.get_settings(config_uri, 'tests')
+    settings = {}
+    config.init(os.environ.get('GEOMAPFISH_CONFIG', '/etc/qgisserver/geomapfish.yaml'))
+    settings.update(config.get_config())
+    return settings
+
+
+def raise_db_error(_):
+    raise DBAPIError('this is a test !', None, None)
```
[the remark this answers is not in the document]
Comment (arnaud-morvan) [reply docker/qgisserver/acceptance_tests/functional/conftest.py]: This function was a copy paste with no use, completely removed.

File: docker/build/requirements.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (sbrunner) [reply docker/build/requirements.txt]: For me it should be simplified in the future, but go ahead like it.

File: docker/qgisserver/geomapfish_qgisserver/accesscontrol.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arnaud-morvan) [reply docker/qgisserver/geomapfish_qgisserver/accesscontrol.py]: Ok for the distinct.
Note it make sense to return a set, it is not ordered, else we need to convert it to a set in tests for comparison with expected set of roles.

[the remark this answers is not in the document]
Comment (sbrunner) [reply docker/qgisserver/geomapfish_qgisserver/accesscontrol.py]: +1 :-)

[the remark this answers is not in the document]
Comment (arnaud-morvan) [reply docker/qgisserver/geomapfish_qgisserver/accesscontrol.py]: You were right, it is not really here that we need a set, as we've only one user_id.
But in get_restriction_areas as we can have many roles with the same restriction_area, unneeded  to make an union with duplicated shapes.
Updated

**Comment by arnaud-morvan** [comment]
@sbrunner : I see that when broadcast is received, self.layers si deleted, but what about sqlalchemy objects stored in sqlalchemy session internal cache. As I understand : 
* Scoped session is supposed to have have the same lifetime as the request
* I agree that we can keep session until we receive invalidate broadcast, but in such case we should remove all thread's sessions on invalidate event.

File: docker/build/requirements.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (arnaud-morvan) [reply docker/build/requirements.txt]: I've adopted a new strategy, as host docker socket is not binded in docker-compose-run, I do not call `make -C docker/qgisserver tests` in root Makefile tests target.
Instead I run a parallel job in Jenkinsfile.

**Comment by arnaud-morvan** [comment]
@sbrunner: All checks and tests are green, missing an approve.
Can you have a last look, hoping to merge before demo upgrade.

**Comment by sbrunner** [comment]
>  ... we should remove all thread's sessions on invalidate event.

You know how to do that easily?

**Comment by arnaud-morvan** [comment]
I'm not sure how we should handle session properly, we might discuss it by phone or open a ticket.

**Pull request merged**